vowpalwabbit/global_data.cc

```diff
@@ -257,17 +260,47 @@ std::string dump_weights_to_json_weight_typed(
         rapidjson::Value offset_value(static_cast<uint64_t>(info.offset != 0 ? info.offset >> info.stride_shift : 0));
         parameter_object.AddMember("offset", offset_value, allocator);
       }
-      else
+      else if (include_feature_names)
       {
         // There is no reverse mapping. We leave nulls in place of terms and offset.
         parameter_object.AddMember("terms", rapidjson::Value(rapidjson::Type::kNullType), allocator);
         parameter_object.AddMember("offset", rapidjson::Value(rapidjson::Type::kNullType), allocator);
       }
 
+      // If include_feature_names is false, we don't add terms or offset
+
       rapidjson::Value index_value(static_cast<uint64_t>(idx));
       parameter_object.AddMember("index", index_value, allocator);
       rapidjson::Value value_value(static_cast<float>(*v));
       parameter_object.AddMember("value", value_value, allocator);
+
+      const float* current_weight_state = &(*v);
+      if (include_online_state)
+      {
+        rapidjson::Value extra_state_value(rapidjson::kObjectType);
+        rapidjson::Value adaptive_value(rapidjson::kNumberType);
+        rapidjson::Value normalized_value(rapidjson::kNumberType);
+
+        if (parameter_holder.adaptive && !parameter_holder.normalized)
+        {
+          adaptive_value = current_weight_state[1];
+          normalized_value = rapidjson::kNullType;
+        }
+        if (!parameter_holder.adaptive && parameter_holder.normalized)
+        {
+          adaptive_value = rapidjson::kNullType;
+          normalized_value = current_weight_state[1];
+        }
+        if (parameter_holder.adaptive && parameter_holder.normalized)
+        {
+          adaptive_value = current_weight_state[1];
+          normalized_value = current_weight_state[2];
+        }
+
+        extra_state_value.AddMember("adaptive", adaptive_value, allocator);
```

Comment:
is it possible to make gd.cc responsible for adding this since this is gd-specific?

Reply:
Yeah I agree it would be good. I am not sure of a good way to make that happen at the moment to be honest.

Reply:
I see - looks like the fact that we have adaptive/normalized field as array_parameters members is hack as well.
Does it mean that if we try to output state with non-gd learner, we just not going to see extra output for now?

Reply:
The dump_json_weights feature explicitly restricts usage to GD for now. I figured it was better to restrict usage until we work out a way to better support other use cases.